lmod_manager/lmod_manager.py

```diff
@@ -7,124 +7,270 @@
+
+
+class Tool:
+    def __init__(self, archive: Path = None, module_name: str = None) -> None:
+        assert (archive and not module_name) or (not archive and module_name)
+
+        if archive:
+            match = re.match(
+                rf"{self.archive_name()}-([\d.w]*(?:-\d*)?)-([\w-]*)-(linux(?:64|)?)-bin\.tar\.gz",
+                archive.name,
+            )
+
+            if not match:
+                raise Error("unexpected archive name format")
+
+            self._archive = archive
+            self._version = match.group(1)
+            self._target = match.group(2)
+            self._linux = match.group(3)
+
+        if module_name:
+            match = re.match(rf"{self.name()}((?:-[^/]*)?)/([\d.w]*(?:-\d*)?)", module_name)
+
+            if not match:
+                raise Error("unexpected module name format")
+
+            self._target = match.group(1)[1:]
+            self._version = match.group(2)
+
+    @staticmethod
+    def from_archive(archive: Path) -> Tool:
+        if archive.name.startswith(Gnat.archive_name()):
+            return Gnat(archive)
+        if archive.name.startswith(Spark.archive_name()):
+            return Spark(archive)
+        if archive.name.startswith(CodePeer.archive_name()):
+            return CodePeer(archive)
+        raise Error("unexpected archive type")
+
+    @staticmethod
+    def from_module(module_name: str) -> Tool:
+        if module_name.startswith(Gnat.name()):
+            return Gnat(module_name=module_name)
+        if module_name.startswith(Spark.name()):
+            return Spark(module_name=module_name)
+        if module_name.startswith(CodePeer.name()):
+            return CodePeer(module_name=module_name)
+        raise Error("unexpected module type")
+
+    @staticmethod
+    @abstractmethod
+    def name() -> str:
+        raise NotImplementedError
+
+    @staticmethod
+    @abstractmethod
+    def archive_name() -> str:
+        raise NotImplementedError
+
+    @property
+    def module_name(self) -> str:
+        if self._target and self._target != "x86_64":
+            return f"{self.name()}-{self._target}"
+        return self.name()
+
+    def install(self, installation_dir: Path, lmod_modules_dir: Path) -> None:
+        installation_dir = self._installation_dir(installation_dir)
+        installation_dir.parent.mkdir(exist_ok=True)
+        config_dir = self._config_dir(lmod_modules_dir)
+        config_dir.mkdir(exist_ok=True)
+        config_file = self._config_file(lmod_modules_dir)
+        config_file.touch()
+
+        shutil.unpack_archive(self._archive, format="gztar")
+        self._install_archive(installation_dir)
+        shutil.rmtree(self._extracted_archive_dir())
+
+        with open(config_file, "w", encoding="utf-8") as f:
+            f.write(
+                f"""\
+local pkgName = myModuleName()
+local version = myModuleVersion()
+local pkg     = pathJoin("{installation_dir}",pkgName,version,"bin")
+prepend_path("PATH", pkg)
+"""
```

Comment:
You could use [`testwrap.dedent`](https://docs.python.org/3.6/library/textwrap.html#textwrap.dedent) to make this a bit more readable.

Reply:
Changed.